SymPy 1.2 release

asmeurer opened the pull request:
This is the release branch for 1.2. This shouldn't be merged until the final release is made. 

Any changes for the release should be made against the `1.2` branch. Only critical changes should be made against this branch.

bjodah committed:
New classes in .codegen.ast

While, Scope, PrintStatement, FunctionPrototype, FunctionDefinition, FunctionCall

bjodah committed:
Fix tests

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Add submodule .utilities._compilation for internal use

bjodah committed:
Add .codegen.algorithms with Newton's method

bjodah committed:
Add .utilities._compilation.util.capture_stdout_stderr

bjodah committed:
Tweak .codegen.ast.Token compatibility with Basic

bjodah committed:
Add new ast nodes to test_ccode.py

bjodah committed:
Handle attributes through .codegen.ast.Node

bjodah committed:
Refactor Nodes in .codegen.ast

bjodah committed:
More .codegen.ast nodes

bjodah committed:
Better str/repr printing of .codegen.ast nodes

bjodah committed:
Initial support for pretty printing codegen ASTs

bjodah committed:
Revert CodeBlock.body -> CodeBlock.args, fix indented printing

bjodah committed:
Update algorithms since update of CodeBlock

bjodah committed:
Work toward enabling fortran code printers for new ast nodes

bjodah committed:
Support for new AST nodes in Python & Fortran

bjodah committed:
Intermediate commit

bjodah committed:
Fix test failures for temporary commit

bjodah committed:
Test fixes, Variable.as_Declaration

bjodah committed:
Fix test failures

bjodah committed:
Enhanced documentation (and minor tweaks)

bjodah committed:
Enhanced documentation, some additions to .codegen.fnodes

bjodah committed:
Fix missing classes to test_args, fix doc warnings

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Fix merge

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Address review from @smichr

bjodah committed:
Drop _print_Integer from mpmath printer

bjodah committed:
Minor improvements to AST nodes, added test_applications.

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Support lt, le, ge, gt on .codegen.ast.Variable

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Settle for only passing kwargs generally in _print_* methods

stewpend0us committed:
wrote a failing test for CodeBlock.cse(). I could have the assertion wrong but the .cse() method actually fails to run. I tracked it down to cse_main.py -> _rebuild(expr): -> if iterable(expr): when expr is a tuple of (more than 1) Matrix the _rebuild method doesn't correctly rebuild the elements of the tuple. Matrix class is expecting (Matrix([..]),) as it's arguments but it's getting Matrix([..]) (missing the sinlge element tuple) I don't konw if that's a problem with the matrix stuff or the cse stuff....but either way :(

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Added doctest to .codegen.algorithms.newtons_method_function

bjodah committed:
Make test for gh-14118 xfail for now

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

stewpend0us committed:
couple of comments that would have helped me. Also some tests in the printer for Element

bjodah committed:
Merge pull request #4 from stewpend0us/codegen-ast-more-nodes

couple of comments that would have helped me

bjodah committed:
fix codegen.rst

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Drop stewpend0us tests for now

bjodah committed:
Update codegen.rst

stewpend0us committed:
some low hanging fruit...

stewpend0us committed:
added missing test

bjodah committed:
Merge pull request #7 from stewpend0us/codegen-ast-more-nodes

some low hanging fruit...

bjodah committed:
Remove spurious __new__ (fixes bjodah/sympy#8)

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

asmeurer committed:
Improve the performance of Basic.__eq__ in Python 2

In PR #12945, Basic.__eq__ was changed to compare types with != instead of "is
not". This lead to a performance degradation in Python 2 because of the
BasicMeta.__cmp__ method (the Basic ordering of classes, see #4269). In Python
2, != calls __cmp__ if __ne__ is not defined. This does not affect Python 3
because __cmp__ is ignored in Python 3.

Removing the ordering of classes has proved to be a more difficult task than I
anticipated (#13059), so for now, we add a workaround. In Python 2, we only
look at type(self) != type(other) if the metaclass of self actually defines
__ne__ (bypassing checks to __cmp__).

This should be compared against the following (incorrect) patch

diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index bdfebf24cc..4b3cf77ee7 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -324,7 +324,7 @@ def __eq__(self, other):
             except SympifyError:
                 return NotImplemented

-            if type(self) != type(other):
+            if type(self) is not type(other):
                 return False

         return self._hashable_content() == other._hashable_content()

Using the benchmark ask(Q.real(x**(y/z)), Q.real(x) & Q.real(y/z) &
Q.rational(y/z) & Q.even(z) & Q.positive(x)) (from #12898), in Python 2:

Master (commit before this):

    In [1]: %timeit ask(Q.real(x**(y/z)), Q.real(x) & Q.real(y/z) & Q.rational(y/z) & Q.even(z) & Q.positive(x))
    1 loop, best of 3: 4.57 s per loop

The above (incorrect) patch:

    In [1]: %timeit ask(Q.real(x**(y/z)), Q.real(x) & Q.real(y/z) & Q.rational(y/z) & Q.even(z) & Q.positive(x))
    1 loop, best of 3: 3.76 s per loop

This commit:

    In [1]: %timeit ask(Q.real(x**(y/z)), Q.real(x) & Q.real(y/z) & Q.rational(y/z) & Q.even(z) & Q.positive(x))
    1 loop, best of 3: 3.85 s per loop

So it isn't quite as fast, but it's about as fast as we can get while
maintaining the proper metaclass __eq__ checks (which are needed for the
Function('f', real=True) feature to work).

Also worth noting that Python 3 master is faster still:

    In [1]: %timeit ask(Q.real(x**(y/z)), Q.real(x) & Q.real(y/z) & Q.rational(y/z) & Q.even(z) & Q.positive(x))
    3.07 s ± 72.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Additionally, it should be pointed out that benchmarks that make direct use of
undefined functions may still be slower than before #12945, due to the more
advanced logic that was added there (I haven't looked into whether
UndefinedFunction.__eq__ itself could be sped up).

asmeurer committed:
Hide warnings printed during the multipledispatch tests

Fixes #14684.

asmeurer committed:
Remove tensorflow from the top-level import

Otherwise tensorflow gets imported along with sympy when it is installed.

asmeurer committed:
Add bin/test_external_imports.py

This script will make sure no external libraries are being imported with 'from
sympy import *' that aren't supposed to be. This will avoid issues in the
future like the one fixed in the previous commit where tensorflow was imported
alongside SymPy.

asmeurer committed:
Run bin/test_external_imports.py in Travis

It has to be run in the optional dependency build because the script only
detects imports if the library is installed.

asmeurer committed:
Fix external imports test in Travis

asmeurer committed:
Use a title instead of a legend for the plotting test

This avoids an error printed to the terminal, and the title is the more
correct thing to do here anyway.

asmeurer committed:
Use os.path.normcase instead of sys_normcase in runtests

For whatever reason, os.path.normcase returns the string unchanged on macOS,
even though the filesystem is case insensitive.

This causes inspect.getmodule to fail on the test files when the tests are
running, because it uses sys.normcase and doesn't recognize the lowercase path
as the same as the normal path, which in turn causes a test I am writing (from
https://github.com/sympy/sympy/pull/12524) to fail.

asmeurer committed:
Print S.true and S.false as just "true" and "false"

Not only is this better, since those names are included in the default
namespace, this is necessary to keep the str printer working with sympify with
safe parsing (https://github.com/sympy/sympy/pull/12524), which disables
attribute access.

asmeurer committed:
Make ordinals objects not singletons

This code hasn't been in a release yet, so there are no compatibility issues.
In general, things should not be put on the singleton registry unless they are
going to be used often and the ability to do is comparison or the memory
savings are worth it. It is not necessary to make an object the only instance
of its class, since == works just fine with two instances. For example (after
this change):

>>> OrdinalOmega() == OrdinalOmega() == omega
True

asmeurer committed:
Add Naturals, Naturals0, and UniversalSet to the 'from sympy import *' namespace

Also added tests that all singleton classes are imported (with exceptions for
rational numbers and NegativeInfinity, which can be accessed with -oo). The
existing singleton test was moved to the new test_singleton.py file.

asmeurer committed:
Add a comment about why the singleton names in namespace test is there

asmeurer committed:
Remove unused imports

asmeurer committed:
Fix doctesting of methods wrapped by fastcache

See also
https://github.com/python/cpython/commit/b532df62b95920c380190f4a468fe15ad0c7d39c.

asmeurer committed:
Remove executable bit from files in sympy/vector/

asmeurer committed:
Fix doctests

Backports a check from the doctest module that is necessary with the change to
inspect.isroutine.

asmeurer committed:
Use inspect.isfunction instead of inspect.isroutine in doctest

isroutine causes the _from_module check to fail in Python 2. I don't think the
difference matters for any functions that we doctest, though I could be wrong
on that.

asmeurer committed:
Remove "S." from singleton set str printers

I haven't fixed the (doc)tests yet.

asmeurer committed:
Check 'register' explicitly in the Singleton tests

SingletonRegistry sets the singletons as attributes, so we cannot
differentiate between methods of the class and singletons automatically.

asmeurer committed:
Test against singletons that haven't been installed yet

asmeurer committed:
Add an exception for the other singleton from the test above

asmeurer committed:
Allow the singleton class to be imported with 'from sympy import *'

This applies to S.EmptySequence, which prints as EmptySequence().

asmeurer committed:
Add 'Integers' to 'from sympy import *'

asmeurer committed:
Fix the printing tests for the fancy set singletons not printing with "S."

asmeurer committed:
Fix doctests for fancyset singletons printing without "S."

asmeurer committed:
Allow singletons in the physics module to not be imported with 'from sympy import *'

asmeurer committed:
Consolidate the optional dependency tests in Travis

asmeurer committed:
Fix test calling syntax

asmeurer committed:
Fix doctest call

asmeurer committed:
Clean up the LaTeX parsing tests

- Don't print stuff in the tests
- Use XFAIL for the failing test

asmeurer committed:
Only print stuff from the latex build_parser function when SYMPY_DEBUG is set

asmeurer committed:
Change collections ABC classes to explicitly use collections.abc

In the upcoming Python 3.7, importing an ABC from collections rather than
collections.abc raises a deprecation warning, which causes the tests to fail.

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

jlumpe committed:
Account for free symbols in codegen.ast.CodeBlock

jlumpe committed:
Check assignment types in CodeBlock.topological_sort

bjodah committed:
Merge pull request #9 from jlumpe/codeblock-free-symbols

Account for free symbols in codegen.ast.CodeBlock

jlumpe committed:
Fix test for CodeBlock.topological_sort

jlumpe committed:
Avoid conflicts between exisiting and auto-generated symbols in CodeBlock.cse()

bjodah committed:
Merge pull request #12 from jlumpe/codeblock-cse-unique-symbols

 Avoid conflicts between exisiting and auto-generated symbols in CodeBlock.cse()

bjodah committed:
Merge pull request #11 from jlumpe/fix-codeblock-toposort-test

Fix test for CodeBlock.topological_sort

jlumpe committed:
Clean up ast.Token code and add docstrings

bjodah committed:
Merge pull request #14 from jlumpe/cleanup-ast-token

Cleanup ast.Token

asmeurer committed:
Add collections.abc imports to compatibility

Python 3.7 makes importing them directly from collections raise a deprecation
warning, but in Python 2.7, they are only available from collections.

asmeurer committed:
Ignore logger messages when importing theano in the theanocode tests

asmeurer committed:
Deactivate the Travis virtualenv when using conda

This should remove the IPython warning about being in a virtualenv in the
tests.

asmeurer committed:
Merge branch 'master' into test-cruft

asmeurer committed:
Install and test matchpy in the optional tests build on Travis

asmeurer committed:
Move deactivate call up so we don't install conda in the virtualenv

asmeurer committed:
Automatically close the plot when unset_show() is used

This way they are closed in the doctests and there is no warning from
matplotlib about too many open plots.

asmeurer committed:
Remove print statements from the rubi submodule

asmeurer committed:
Run Rubi doctests on Travis in the optional dependencies build

asmeurer committed:
Hide tensorflow warnings in the lambdify tests

asmeurer committed:
Fix reporting of certain errors in the test runner

asmeurer committed:
Move the deactivate call outside the if block

asmeurer committed:
Use set -x in .travis.yml

asmeurer committed:
Fix error building the LaTeX parser

The cwd is set to the temporary output directory, so the grammar file needs to
use an absolute path.

asmeurer committed:
Handle the file not existing in the plot test cleanup

asmeurer committed:
Install Sage with everything else in Travis

Otherwise it will downgrade matplotlib, causing some tests to fail.

asmeurer committed:
In Travis, test Sage in its own conda environment

asmeurer committed:
Remove set -x from .travis.yml

It's too verbose and Travis has a hard time parsing it.

asmeurer committed:
Fix TEST_SAGE check in .travis.yml

asmeurer committed:
Update version for 1.2rc1

asmeurer committed:
Pull first so the release branch is fetched

asmeurer committed:
Update .mailmap

asmeurer committed:
Update AUTHORS

asmeurer committed:
Run the Sphinx doctests in the optional dependency Travis build

asmeurer committed:
Fix testing of Sphinx doctests in the optional dependency Travis build

sidhantnagpal commented on .mailmap:
```diff
@@ -382,3 +382,10 @@ Ravi charan <ravicharan.vsp@gmail.com>
+Rahil Hastu <rahilhastu@gmail.com> rahil hastu <rahilhastu@gmail.com>
+Sidhant Nagpal <sidhantnagpal97@gmail.com> sidhantnagpal <sidhantnagpal97@gmail.com>
+Sidhant Nagpal <sidhantnagpal97@gmail.com> “sidhantnagpal” <sidhantnagpal97@gmail.com>
+Sidhant Nagpal <sidhantnagpal97@gmail.com> Sidhant Nagpal <36465988+sidhantnagpal@users.noreply.github.com>
```
Is this file autogenerated?
I would just want `Sidhant Nagpal <sidhantnagpal97@gmail.com>`, just like in AUTHORS.

asmeurer replied:
If the version in AUTHORS is the one you want then this file is correct. See the comments at the top of the file.

asmeurer committed:
Merge pull request #14747 from asmeurer/test-cruft

Clean up some cruft in the tests

asmeurer committed:
Only doctest the numeric computation Sphinx file in the optional dependency build

The build is starting to time out on Travis, so we need to cut it down.

asmeurer committed:
Merge branch '1.2' into core-py2-performance

asmeurer committed:
Add 3.7 testing to the release script

It isn't enabled yet because 3.7 isn't actually out yet.

asmeurer committed:
Test isympy in the tarball test

asmeurer committed:
Make sure bin/isympy is included in the tarball, even when setup.py uses setuptools

asmeurer committed:
Update release tarball whitelist

asmeurer committed:
Update release tarball whitelist

asmeurer committed:
Include rubi header template in installed files

asmeurer committed:
Merge pull request #14685 from asmeurer/core-py2-performance

Improve the performance of Basic.__eq__ in Python 2

asmeurer committed:
Remove executable bit from sympy/vector/tests/test_dyadic.py

bjodah committed:
Merge branch 'master' into codegen-ast-more-nodes

bjodah committed:
Merge branch '1.2' into codegen-ast-more-nodes

asmeurer committed:
Remove sage from the TEST_OPT_DEPENDENCY variable

The sage tests are run in a separate environment. See https://github.com/sympy/sympy/pull/14747.

asmeurer committed:
Merge pull request #14844 from asmeurer/safe-sympify-backport

Safe sympify backport

asmeurer committed:
Indicate the protocol in the pickling assertion error

asmeurer committed:
Handle case insensitive filesystems properly in the LaTeX parser generation

The glob module does not match case insensitively on case insensitive systems.
Furthermore, the logic of the function assumed that LaTeXparser.py and
latexparser.py would be different files, but they aren't on such systems.

asmeurer committed:
Maintain the encoding in the LaTeX parser files

The default header from antlr is removed and replaced with a custom header
because it has a full filesystem path to the grammar file, but the encoding
was part of the removed antlr header as well. This puts it in the custom
header.

bjodah committed:
Clean-up after test compilation

bjodah committed:
Add cython & wurlitzer to opt. conda deps.

bjodah committed:
Merge branch '1.2' into codegen-ast-more-nodes

bjodah committed:
CI: pip install wurlitzer, set -e, add codegen to test list

bjodah committed:
Remove **kwargs from Printer._print_Class

bjodah committed:
Fix doctest of indented printing

bjodah committed:
revert changes to .printing.theanocode

bjodah committed:
Use TemporaryDirectory contex manager in tests

bjodah committed:
Use importlib in Py3

bjodah committed:
Set compiler environment variables for conda env

bjodah committed:
wurlitzer from conda-forge

bjodah committed:
CI: Require compilers to work in specific Travis runs

bjodah committed:
CI: set LD_LIBRARY_PATH

bjodah committed:
More specific exception for missing compiler

asmeurer committed:
Test runner: print the full exception chain in Python 3

For example:

    def raise_valueerror():
        raise ValueError

    def test_chained_exception():
        try:
            raise_valueerror()
        except ValueError:
            raise TypeError

used to show

    ________________________________________________________________________________
    __________ sympy/utilities/tests/test_test.py:test_chained_exception ___________
    Traceback (most recent call last):
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/tests/test_test.py", line 8, in test_chained_exception
        raise TypeError
    TypeError

Now it shows

    ________________________________________________________________________________
    __________ sympy/utilities/tests/test_test.py:test_chained_exception ___________
    Traceback (most recent call last):
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/tests/test_test.py", line 6, in test_chained_exception
        raise_valueerror()
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/tests/test_test.py", line 2, in raise_valueerror
        raise ValueError
    ValueError

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/tests/test_test.py", line 8, in test_chained_exception
        raise TypeError
    TypeError

bjodah committed:
Fixes to compilation

- may_xfail on macOS & Windows
- Fix link_py_so for python built without shared library
- More fine grained skips in tests of codegen

asmeurer committed:
Merge pull request #13100 from bjodah/codegen-ast-more-nodes

Codegen: More AST nodes

asmeurer committed:
Update AUTHORS

asmeurer committed:
Add more optional dependencies to the release script

asmeurer committed:
Run the tarball test against Python 3.7

asmeurer committed:
Fix some text in the release README

asmeurer committed:
Fix the GitHub release text

The files are now shown above the text, not below.

asmeurer committed:
Bump the version to 1.2rc2

jlumpe committed:
Remove ast.Statement class

jlumpe committed:
CodeBlock elements printed as statements in C

asmeurer committed:
Merge pull request #14867 from jlumpe/remove-ast-statement

Remove ast.Statement

asmeurer committed:
Bump version to 1.2 (final)

asmeurer commented:
OK, the release is out. Once Travis finishes building this should be merged into master and the 1.2 branch deleted.
